pkg/controller/deployment/deployment_controller.go

```diff
@@ -306,6 +317,50 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	dc.queue.Forget(key)
 }
 
+// classifyReplicaSets uses NewReplicaSetControllerRefManager to classify ReplicaSets
+// and adopts them if their labels match the Deployment but are missing the reference.
+// It also removes the controllerRef for ReplicaSets, whose labels no longer matches
+// the deployment.
+func (dc *DeploymentController) classifyReplicaSets(deployment *extensions.Deployment) error {
+	rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Error getting ReplicaSets for deployment %q: %v", deployment.Name, err)
+		return err
+	}
+
+	deploymentSelector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("deployment %s/%s has invalid label selector: %v", deployment.Namespace, deployment.Name, err)
+	}
+	cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, deployment.ObjectMeta, deploymentSelector, getDeploymentKind())
+	matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(rsList)
+	for _, replicaSet := range matchesNeedsController {
+		err := cm.AdoptReplicaSet(replicaSet)
+		// continue to next RS if adoption fails.
+		if err != nil {
+			// If the RS no longer exists, don't even log the error.
+			if !errors.IsNotFound(err) {
+				utilruntime.HandleError(err)
+			}
+		} else {
+			matchesAndControlled = append(matchesAndControlled, replicaSet)
+		}
+	}
+	// remove the controllerRef for the RS that no longer have matching labels
+	var errlist []error
+	for _, replicaSet := range controlledDoesNotMatch {
```

Comment:
Yes, that's the suggested approach but we have had clients who don't want to use a second Deployment and instead relabel pods. Changing the selector for a service makes sense, not sure how much does it make for a Deployment but I don't want to rule anything out especially once we have custom Deployments.